lib/wallet/test/unit/Cardano/Wallet/DB/Store/Delegations/Migrations/V5Spec.hs

```diff
@@ -0,0 +1,170 @@
+    )
+import Control.Monad.Logger
+    ( NoLoggingT (..)
+    )
+import Control.Monad.Reader
+    ( ReaderT (..)
+    )
+import Control.Tracer
+    ( nullTracer
+    )
+import Database.Persist.Sql
+    ( Entity (..)
+    , PersistStoreWrite (..)
+    , insertMany_
+    , selectList
+    )
+import Database.Persist.Sqlite
+    ( withSqliteConn
+    )
+import Test.Hspec
+    ( Spec
+    , describe
+    , it
+    , shouldBe
+    , shouldNotBe
+    )
+import Test.QuickCheck
+    ( Gen
+    , InfiniteList (..)
+    , Positive (..)
+    , arbitrary
+    , elements
+    , generate
+    , oneof
+    )
+import UnliftIO.Resource
+    ( runResourceT
+    )
+
+import Cardano.Api
+    ( SlotNo
+    )
+import Cardano.Wallet.Primitive.Types.Pool
+    ( PoolId
+    )
+import Data.List
+    ( maximumBy
+    , sortOn
+    )
+import Data.Ord
+    ( Down (..)
+    , comparing
+    )
+
+import qualified Data.ByteString as BS
+import qualified Data.Text as T
+
+spec :: Spec
+spec =
+    describe "Delegations table migration"
+        $ it "'migrate' db new delegation table"
+        $ testMigrationDelegationsTable
+            "before_new_delegation.sqlite"
+
+testMigrationDelegationsTable :: FilePath -> IO ()
+testMigrationDelegationsTable dbName = do
+    let performMigrations path =
+            runMigrations
+                (newMigrationInterface nullTracer)
+                path
+                newStyleMigrations
+        testOnCopiedAndMigrated test = fmap snd
+            $ withinCopiedFile dbName
+            $ \path _ -> do
+                performMigrations path
+                test path
+    testOnCopiedAndMigrated persistentTableIsCompatible
+  where
+    persistentTableIsCompatible :: FilePath -> IO ()
+    persistentTableIsCompatible path = do
+        runResourceT . runNoLoggingT
+            $ withSqliteConn (T.pack path)
+            $ runReaderT
+            $ do
+                -- delegations from the migration
+                oldDelegations <-
+                    fmap entityVal
+                        <$> selectList [] []
+                liftIO $ length oldDelegations `shouldNotBe` 0
+                -- last used slot
+                let latest = maximumBy (comparing delegationSlot) oldDelegations
+                -- 10 delegations to insert
+                delegations <-
+                    sortOn delegationOrd
+                        <$> liftIO (someDelegations 10 $ delegationSlot latest)
+                insertMany_ delegations
+                -- last 10 delegations
+                delegations' <-
+                    reverse
+                        . take 10
+                        . sortOn (Down . delegationOrd)
+                        . fmap entityVal
+                        <$> selectList [] []
+                liftIO $ delegations' `shouldBe` delegations
+
+-- generate n delegations with a slot number greater than the given slot and
+-- all different slots
+generateDelegations :: Int -> SlotNo -> Gen [Delegations]
```

Comment:
(Just for informatio:

This function is fine, but a "wholemeal" approach to writing this function would be

```hs
generateDelegations n lslot = do
    deltas <- vectorOf n (arbitrary :: Gen Positive)
    let slots = scanl (+) lslot deltas
    vectorOf n (zipWith (\f (Positive slot) -> f slot) generateDelegation slots)

generateDelegation :: Slot → Gen Delegation
generateDelegation = … -- generates a single delegation
```

Here,

* primitive recursion is avoided
* the computation of the slots is decoupled from the generation of the rows

But the current definition is fine, too.
 
)